{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7634bcba-6c9a-4805-97c6-e53e408a175b",
   "metadata": {},
   "source": [
    "## Genome quality control and selection\n",
    "All candidate genomes were evaluated using a quality control procedure based on assembly metrics. For each assembly, a composite quality score (`quality_score`) was calculated by integrating:\n",
    "- Assembly continuity (N50)\n",
    "- Sequencing depth (coverage)\n",
    "- Assembly fragmentation (number of contigs â‰¥200 bp)\n",
    "\n",
    "Genomes were then grouped by geographic region. Within each region, the five assemblies with the highest quality scores were selected. This stratified sampling strategy was used to:\n",
    "- Maximize the overall quality of the final dataset\n",
    "- Preserve regional representativeness\n",
    "- Avoid bias caused by over-representation of regions with higher numbers of available isolates\n",
    "\n",
    "Only the top-ranked genomes per region, together with their associated metadata and quality metrics, were retained for downstream analyses."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1ea49567-d8d6-442f-900b-03dba9b40a5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "from pathlib import Path\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "dddd1277-8ca6-435a-b869-583f085d2774",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_df = Path('/salmonella-pangenome-SP/data/Metadata/meta_sp_total.txt')\n",
    "assert path_df.exists(), f'no se encuentra el fichero'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8f4d12eb-5cca-4004-adf5-0db6e6757c16",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Uberstrain        Name  \\\n",
      "0  SAL_BA6133AA   OSF031120   \n",
      "1  SAL_DA9157AA  EC20100134   \n",
      "2  SAL_DA9158AA  EC20100130   \n",
      "\n",
      "  Data Source(Accession No.;Sequencing Platform;Sequencing Library;Insert Size;Experiment;Bases;Average Length;Status)  \\\n",
      "0  SRR1995504;ILLUMINA;Paired;500;SRX1008539;2371...                                                                     \n",
      "1  SRR1183994;LS454;Single;;SRX481517;;;NEW,GCF_0...                                                                     \n",
      "2  SRR1183993;LS454;Single;;SRX481516;;;NEW,GCF_0...                                                                     \n",
      "\n",
      "        Barcode Source Niche Source Type Source Details Collection Year  \\\n",
      "0  SAL_BA6133AA      Poultry     Poultry        chicken            2014   \n",
      "1  SAL_DA9157AA      Poultry     Poultry        chicken            2009   \n",
      "2  SAL_DA9158AA      Poultry     Poultry        chicken            2009   \n",
      "\n",
      "  Collection Month Collection Day  ... Release Date     Status Coverage  \\\n",
      "0                4            NaN  ...   2015-04-24  Assembled       51   \n",
      "1              NaN            NaN  ...   2015-03-05  Assembled        1   \n",
      "2              NaN            NaN  ...   2015-03-05  Assembled        1   \n",
      "\n",
      "       N50   Length                     Species Contig Number(>=200 bp)  \\\n",
      "0    89072  4733951  Salmonella enterica;100.0%                     130   \n",
      "1  4766248  4766248                         NaN                       1   \n",
      "2  4773474  4773474                         NaN                       1   \n",
      "\n",
      "  Low Quality Bases Version Assembly Barcode  \n",
      "0             19814     2.2  SAL_EA5295AA_AS  \n",
      "1                 8     NaN  SAL_KA3434AA_AS  \n",
      "2                15     NaN  SAL_KA3433AA_AS  \n",
      "\n",
      "[3 rows x 41 columns]\n",
      "original columns ['Uberstrain', 'Name', 'Data Source(Accession No.;Sequencing Platform;Sequencing Library;Insert Size;Experiment;Bases;Average Length;Status)', 'Barcode', 'Source Niche', 'Source Type', 'Source Details', 'Collection Year', 'Collection Month', 'Collection Day', 'Collection Time', 'Continent', 'Country', 'Region', 'District', 'City', 'Post Code', 'Latitude', 'Longitude', 'Serovar', 'Subspecies', 'Disease', 'Antigenic Formulas', 'Lab Contact', 'Phage Type', 'Comment', 'Bio Project ID', 'Project ID', 'Sample ID', 'Secondary Sample ID', 'Date Entered', 'Release Date', 'Status', 'Coverage', 'N50', 'Length', 'Species', 'Contig Number(>=200 bp)', 'Low Quality Bases', 'Version', 'Assembly Barcode']\n",
      "Shape:  (175, 41)\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(path_df, sep='\\t', dtype=str, engine='python')\n",
    "print(df.head(3))\n",
    "print('original columns', list(df.columns))\n",
    "print('Shape: ', df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f647aa5-8a71-4ded-8d9a-caeac70801e3",
   "metadata": {},
   "source": [
    "## Normalize columns downloaded from Enterobase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "71343667-e508-4f25-9ed7-63d825e8ab13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Normalize columns: ['uberstrain', 'name', 'data_source_accession_no_sequencing_platform_sequencing_library_insert_size_experiment_bases_average_length_status', 'barcode', 'source_niche', 'source_type', 'source_details', 'collection_year', 'collection_month', 'collection_day', 'collection_time', 'continent', 'country', 'region', 'district', 'city', 'post_code', 'latitude', 'longitude', 'serovar', 'subspecies', 'disease', 'antigenic_formulas', 'lab_contact', 'phage_type', 'comment', 'bio_project_id', 'project_id', 'sample_id', 'secondary_sample_id', 'date_entered', 'release_date', 'status', 'coverage', 'n50', 'length', 'species', 'contig_number', 'low_quality_bases', 'version', 'assembly_barcode']\n"
     ]
    }
   ],
   "source": [
    "def normalize_col(col: str) -> str:\n",
    "    c = str(col).strip()    \n",
    "    c = re.sub(r\"[^A-Za-z0-9]+\", \"_\", c)\n",
    "    c = c.replace(\"_200_bp\", \"\")\n",
    "    c = re.sub(r\"_+\", \"_\", c)                 \n",
    "    return c.strip(\"_\").lower()\n",
    "\n",
    "df.columns = [normalize_col(c) for c in df.columns]\n",
    "print(\"Normalize columns:\", list(df.columns))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b12ac24-b6e3-4e96-96c9-c4f968f51806",
   "metadata": {},
   "source": [
    "## Identify and split composite columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7638ccc6-a934-4d12-aa0b-66814dcb53e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     uberstrain        name       barcode source_niche source_type  \\\n",
      "0  SAL_BA6133AA   OSF031120  SAL_BA6133AA      Poultry     Poultry   \n",
      "1  SAL_DA9157AA  EC20100134  SAL_DA9157AA      Poultry     Poultry   \n",
      "\n",
      "  source_details collection_year collection_month collection_day  \\\n",
      "0        chicken            2014                4            NaN   \n",
      "1        chicken            2009              NaN            NaN   \n",
      "\n",
      "  collection_time  ... low_quality_bases version assembly_barcode  \\\n",
      "0             NaN  ...             19814     2.2  SAL_EA5295AA_AS   \n",
      "1             NaN  ...                 8     NaN  SAL_KA3434AA_AS   \n",
      "\n",
      "  accession_no sequencing_platform sequencing_library insert_size  experiment  \\\n",
      "0   SRR1995504            ILLUMINA             Paired         500  SRX1008539   \n",
      "1   SRR1183994               LS454             Single               SRX481517   \n",
      "\n",
      "       bases average_length  \n",
      "0  237155400            225  \n",
      "1                            \n",
      "\n",
      "[2 rows x 47 columns]\n"
     ]
    }
   ],
   "source": [
    "column_c = [c for c in df.columns if 'data_source_' in c]\n",
    "expanded = df[column_c[0]].str.split(';', n=7, expand=True)\n",
    "expanded.columns = ['accession_no',\n",
    "                    'sequencing_platform',\n",
    "                    'sequencing_library',\n",
    "                    'insert_size',\n",
    "                    'experiment',\n",
    "                    'bases',\n",
    "                    'average_length',\n",
    "                    'status'\n",
    "                   ]  \n",
    "for col in expanded.columns: \n",
    "    df[col] = expanded[col].str.strip() \n",
    "\n",
    "# Remove the original column\n",
    "\n",
    "df = df.drop(columns=[column_c[0]])\n",
    "print(df.head(2))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "45aa338c-9774-453d-9a0d-64d1e55c52b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert the variables of interest to numeric values\n",
    "for col in ['n50', 'coverage', 'contig_number']:\n",
    "    df[col]= pd.to_numeric(df[col], errors='coerce')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ca3bc9a2-15b1-4e2e-8bd8-6d906c2da6a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     uberstrain        name       barcode source_niche source_type  \\\n",
      "0  SAL_BA6133AA   OSF031120  SAL_BA6133AA      Poultry     Poultry   \n",
      "1  SAL_DA9157AA  EC20100134  SAL_DA9157AA      Poultry     Poultry   \n",
      "2  SAL_DA9158AA  EC20100130  SAL_DA9158AA      Poultry     Poultry   \n",
      "3  SAL_DA9159AA  EC20090884  SAL_DA9159AA      Poultry     Poultry   \n",
      "4  SAL_DA9160AA  EC20090531  SAL_DA9160AA      Poultry     Poultry   \n",
      "\n",
      "  source_details collection_year collection_month collection_day  \\\n",
      "0        chicken            2014                4            NaN   \n",
      "1        chicken            2009              NaN            NaN   \n",
      "2        chicken            2009              NaN            NaN   \n",
      "3        chicken            2009              NaN            NaN   \n",
      "4        chicken            2009              NaN            NaN   \n",
      "\n",
      "  collection_time  ... version assembly_barcode accession_no  \\\n",
      "0             NaN  ...     2.2  SAL_EA5295AA_AS   SRR1995504   \n",
      "1             NaN  ...     NaN  SAL_KA3434AA_AS   SRR1183994   \n",
      "2             NaN  ...     NaN  SAL_KA3433AA_AS   SRR1183993   \n",
      "3             NaN  ...     NaN  SAL_KA3474AA_AS   SRR1183992   \n",
      "4             NaN  ...     NaN  SAL_KA3473AA_AS   SRR1183991   \n",
      "\n",
      "  sequencing_platform sequencing_library insert_size  experiment      bases  \\\n",
      "0            ILLUMINA             Paired         500  SRX1008539  237155400   \n",
      "1               LS454             Single               SRX481517              \n",
      "2               LS454             Single               SRX481516              \n",
      "3               LS454             Single               SRX481515              \n",
      "4               LS454             Single               SRX481514              \n",
      "\n",
      "  average_length quality_score  \n",
      "0            225      4.698144  \n",
      "1                    19.502496  \n",
      "2                    19.516948  \n",
      "3                    19.490898  \n",
      "4                    19.439780  \n",
      "\n",
      "[5 rows x 48 columns]\n",
      "     uberstrain        name       barcode source_niche source_type  \\\n",
      "0  SAL_BA6133AA   OSF031120  SAL_BA6133AA      Poultry     Poultry   \n",
      "1  SAL_DA9157AA  EC20100134  SAL_DA9157AA      Poultry     Poultry   \n",
      "2  SAL_DA9158AA  EC20100130  SAL_DA9158AA      Poultry     Poultry   \n",
      "3  SAL_DA9159AA  EC20090884  SAL_DA9159AA      Poultry     Poultry   \n",
      "4  SAL_DA9160AA  EC20090531  SAL_DA9160AA      Poultry     Poultry   \n",
      "\n",
      "  source_details collection_year collection_month collection_day  \\\n",
      "0        chicken            2014                4            NaN   \n",
      "1        chicken            2009              NaN            NaN   \n",
      "2        chicken            2009              NaN            NaN   \n",
      "3        chicken            2009              NaN            NaN   \n",
      "4        chicken            2009              NaN            NaN   \n",
      "\n",
      "  collection_time  ... version assembly_barcode accession_no  \\\n",
      "0             NaN  ...     2.2  SAL_EA5295AA_AS   SRR1995504   \n",
      "1             NaN  ...     NaN  SAL_KA3434AA_AS   SRR1183994   \n",
      "2             NaN  ...     NaN  SAL_KA3433AA_AS   SRR1183993   \n",
      "3             NaN  ...     NaN  SAL_KA3474AA_AS   SRR1183992   \n",
      "4             NaN  ...     NaN  SAL_KA3473AA_AS   SRR1183991   \n",
      "\n",
      "  sequencing_platform sequencing_library insert_size  experiment      bases  \\\n",
      "0            ILLUMINA             Paired         500  SRX1008539  237155400   \n",
      "1               LS454             Single               SRX481517              \n",
      "2               LS454             Single               SRX481516              \n",
      "3               LS454             Single               SRX481515              \n",
      "4               LS454             Single               SRX481514              \n",
      "\n",
      "  average_length quality_score  \n",
      "0            225      4.698144  \n",
      "1                    19.502496  \n",
      "2                    19.516948  \n",
      "3                    19.490898  \n",
      "4                    19.439780  \n",
      "\n",
      "[5 rows x 48 columns]\n"
     ]
    }
   ],
   "source": [
    "# Creating quality metrics\n",
    "df['quality_score']= df['n50'].fillna(0)/500000 + df['coverage'].fillna(0)/50 + (200 - df['contig_number'].fillna(200))/20\n",
    "# It's divided so that each metric has a similar weight. If we don't know how many contigs that entry has, \n",
    "# we penalize it heavily, assuming a gigantic value, so that sample will come out with a very low score and will almost never be the best.\n",
    "\n",
    "print(df.head(5))\n",
    "\n",
    "# Working just with poultry sources\n",
    "df_trabajo = df[df['source_niche']=='Poultry']\n",
    "\n",
    "print(df_trabajo.head(5))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4ac7afd-eeb8-46f2-8a17-b7aa0c9c1c5f",
   "metadata": {},
   "source": [
    "## Working dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "3693fc8c-9e74-48bf-8a0a-ce5c5848021f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              region        name   accession_no assembly_barcode      serovar  \\\n",
      "11           Alberta  EC20090641     SRR1183981  SAL_KA3404AA_AS  Enteritidis   \n",
      "35           Alberta  EC20111515  GCF_000624315  SAL_PA7468AA_AS  Enteritidis   \n",
      "32           Alberta  EC20120677  GCF_000625755  SAL_PA7476AA_AS  Enteritidis   \n",
      "34           Alberta  EC20111514  GCF_000624295  SAL_PA7505AA_AS  Enteritidis   \n",
      "10           Alberta  EC20090698     SRR1183982  SAL_KA3468AA_AS  Enteritidis   \n",
      "31  British Columbia  EC20120765  GCF_000624995  SAL_PA7576AA_AS  Enteritidis   \n",
      "30  British Columbia  EC20111554  GCF_000624335  SAL_PA7506AA_AS  Enteritidis   \n",
      "33  British Columbia  EC20111510  GCF_000626555  SAL_PA7485AA_AS  Enteritidis   \n",
      "17  British Columbia  EC20120686  GCF_000625555  SAL_PA7535AA_AS  Enteritidis   \n",
      "37  British Columbia  EC20120685  GCF_000625535  SAL_PA7470AA_AS  Enteritidis   \n",
      "46     New Brunswick   N13-02934     SRR5241846  SAL_QA6543AA_AS   Heidelberg   \n",
      "68     New Brunswick   N13-02944     SRR5241820  SAL_QA5373AA_AS   Heidelberg   \n",
      "42     New Brunswick   N13-02946     SRR5241852  SAL_QA6290AA_AS   Heidelberg   \n",
      "19       Nova Scotia  EC20120590  GCF_000625495  SAL_PA7597AA_AS  Enteritidis   \n",
      "29           Ontario  EC20120916  GCF_000624155  SAL_KA3476AA_AS  Enteritidis   \n",
      "5            Ontario  EC20090332     SRR1183990  SAL_KA3477AA_AS  Enteritidis   \n",
      "7            Ontario  EC20090193     SRR1183988  SAL_KA3478AA_AS  Enteritidis   \n",
      "2            Ontario  EC20100130     SRR1183993  SAL_KA3433AA_AS  Enteritidis   \n",
      "6            Ontario  EC20090135     SRR1183989  SAL_KA3479AA_AS  Enteritidis   \n",
      "90            Quebec   N13-01312     SRR5239201  SAL_QA5780AA_AS   Heidelberg   \n",
      "82            Quebec   N13-01311     SRR5239213  SAL_QA6400AA_AS   Heidelberg   \n",
      "57            Quebec   N13-01348     SRR5241832  SAL_QA5385AA_AS   Heidelberg   \n",
      "53            Quebec   N13-01330     SRR5241836  SAL_QA5389AA_AS   Heidelberg   \n",
      "51            Quebec   N13-01323     SRR5241839  SAL_QA5815AA_AS   Heidelberg   \n",
      "\n",
      "   sequence_type source_niche                  source_details collection_year  \\\n",
      "11          ST11      Poultry       chicken-abbatoir-Sporadic            2008   \n",
      "35          ST11      Poultry                  ANIMAL-chicken            2010   \n",
      "32          ST11      Poultry                  ANIMAL-chicken            2011   \n",
      "34          ST11      Poultry                  ANIMAL-chicken            2010   \n",
      "10          ST11      Poultry       chicken-abbatoir-Sporadic            2008   \n",
      "31          ST11      Poultry                  ANIMAL-chicken            2011   \n",
      "30          ST11      Poultry                  ANIMAL-chicken            2010   \n",
      "33          ST11      Poultry                  ANIMAL-chicken            2010   \n",
      "17          ST11      Poultry                  ANIMAL-chicken            2011   \n",
      "37          ST11      Poultry                  ANIMAL-chicken            2011   \n",
      "46          ST15      Poultry  Animal cecal content (chicken)            2011   \n",
      "68          ST15      Poultry  Animal cecal content (chicken)            2012   \n",
      "42          ST15      Poultry  Animal cecal content (chicken)            2012   \n",
      "19          ST11      Poultry                  ANIMAL-chicken            2011   \n",
      "29          ST11      Poultry                  ANIMAL-chicken            2012   \n",
      "5           ST11      Poultry                         chicken            2008   \n",
      "7           ST11      Poultry                         chicken            2008   \n",
      "2           ST11      Poultry                         chicken            2009   \n",
      "6           ST11      Poultry                         chicken            2008   \n",
      "90          ST15      Poultry  Animal cecal content (chicken)            2012   \n",
      "82          ST15      Poultry  Animal cecal content (chicken)            2012   \n",
      "57          ST15      Poultry  Animal cecal content (chicken)            2012   \n",
      "53          ST15      Poultry  Animal cecal content (chicken)            2012   \n",
      "51          ST15      Poultry  Animal cecal content (chicken)            2012   \n",
      "\n",
      "   country                                        lab_contact      n50  \\\n",
      "11  Canada  Laboratory for Foodborne Zoonoses, Public Heal...  4712309   \n",
      "35  Canada  Laboratory for Foodborne Zoonoses, Public Heal...  4685847   \n",
      "32  Canada  Laboratory for Foodborne Zoonoses, Public Heal...  4685846   \n",
      "34  Canada  Laboratory for Foodborne Zoonoses, Public Heal...  4685845   \n",
      "10  Canada  Laboratory for Foodborne Zoonoses, Public Heal...  4680996   \n",
      "31  Canada  Laboratory for Foodborne Zoonoses, Public Heal...  4685984   \n",
      "30  Canada  Laboratory for Foodborne Zoonoses, Public Heal...  4685846   \n",
      "33  Canada                        Public Health Agency Canada  4685842   \n",
      "17  Canada  Laboratory for Foodborne Zoonoses, Public Heal...  4685835   \n",
      "37  Canada  Laboratory for Foodborne Zoonoses, Public Heal...  4685833   \n",
      "46  Canada                     Public Health Agency of Canada   693848   \n",
      "68  Canada                     Public Health Agency of Canada   694160   \n",
      "42  Canada                     Public Health Agency of Canada   412159   \n",
      "19  Canada  Laboratory for Foodborne Zoonoses, Public Heal...  4685840   \n",
      "29  Canada                        Public Health Agency Canada  4807994   \n",
      "5   Canada  Laboratory for Foodborne Zoonoses, Public Heal...  4785193   \n",
      "7   Canada  Laboratory for Foodborne Zoonoses, Public Heal...  4774402   \n",
      "2   Canada  Laboratory for Foodborne Zoonoses, Public Heal...  4773474   \n",
      "6   Canada  Laboratory for Foodborne Zoonoses, Public Heal...  4771194   \n",
      "90  Canada                     Public Health Agency of Canada   725299   \n",
      "82  Canada                     Public Health Agency of Canada   725299   \n",
      "57  Canada                     Public Health Agency of Canada   740519   \n",
      "53  Canada                     Public Health Agency of Canada   725307   \n",
      "51  Canada                     Public Health Agency of Canada   439677   \n",
      "\n",
      "    contig_number  coverage  quality_score  \\\n",
      "11              1       1.0      19.394618   \n",
      "35              1       1.0      19.341694   \n",
      "32              1       1.0      19.341692   \n",
      "34              1       1.0      19.341690   \n",
      "10              1       1.0      19.331992   \n",
      "31              1       1.0      19.341968   \n",
      "30              1       1.0      19.341692   \n",
      "33              1       1.0      19.341684   \n",
      "17              1       1.0      19.341670   \n",
      "37              1       1.0      19.341666   \n",
      "46             43     215.0      13.537696   \n",
      "68             46     164.0      12.368320   \n",
      "42             48     129.0      11.004318   \n",
      "19              1       1.0      19.341680   \n",
      "29              1       1.0      19.585988   \n",
      "5               1       1.0      19.540386   \n",
      "7               1       1.0      19.518804   \n",
      "2               1       1.0      19.516948   \n",
      "6               1       1.0      19.512388   \n",
      "90             39     192.0      13.340598   \n",
      "82             36     178.0      13.210598   \n",
      "57             34     155.0      12.881038   \n",
      "53             39     164.0      12.780614   \n",
      "51             37     176.0      12.549354   \n",
      "\n",
      "                                 sample_fasta         fasta_source  \n",
      "11              Alberta_EC20090641_SRR1183981  EnteroBase_assembly  \n",
      "35           Alberta_EC20111515_GCF_000624315          NCBI_RefSeq  \n",
      "32           Alberta_EC20120677_GCF_000625755          NCBI_RefSeq  \n",
      "34           Alberta_EC20111514_GCF_000624295          NCBI_RefSeq  \n",
      "10              Alberta_EC20090698_SRR1183982  EnteroBase_assembly  \n",
      "31  British_Columbia_EC20120765_GCF_000624995          NCBI_RefSeq  \n",
      "30  British_Columbia_EC20111554_GCF_000624335          NCBI_RefSeq  \n",
      "33  British_Columbia_EC20111510_GCF_000626555          NCBI_RefSeq  \n",
      "17  British_Columbia_EC20120686_GCF_000625555          NCBI_RefSeq  \n",
      "37  British_Columbia_EC20120685_GCF_000625535          NCBI_RefSeq  \n",
      "46         New_Brunswick_N13-02934_SRR5241846  EnteroBase_assembly  \n",
      "68         New_Brunswick_N13-02944_SRR5241820  EnteroBase_assembly  \n",
      "42         New_Brunswick_N13-02946_SRR5241852  EnteroBase_assembly  \n",
      "19       Nova_Scotia_EC20120590_GCF_000625495          NCBI_RefSeq  \n",
      "29           Ontario_EC20120916_GCF_000624155          NCBI_RefSeq  \n",
      "5               Ontario_EC20090332_SRR1183990  EnteroBase_assembly  \n",
      "7               Ontario_EC20090193_SRR1183988  EnteroBase_assembly  \n",
      "2               Ontario_EC20100130_SRR1183993  EnteroBase_assembly  \n",
      "6               Ontario_EC20090135_SRR1183989  EnteroBase_assembly  \n",
      "90                Quebec_N13-01312_SRR5239201  EnteroBase_assembly  \n",
      "82                Quebec_N13-01311_SRR5239213  EnteroBase_assembly  \n",
      "57                Quebec_N13-01348_SRR5241832  EnteroBase_assembly  \n",
      "53                Quebec_N13-01330_SRR5241836  EnteroBase_assembly  \n",
      "51                Quebec_N13-01323_SRR5241839  EnteroBase_assembly  \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_5250/1706637055.py:4: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(lambda x: x.nlargest(5, 'quality_score'))\n"
     ]
    }
   ],
   "source": [
    "# Creating metadata with the main data of the genomes to be worked on\n",
    "df_final = (\n",
    "    df_trabajo\n",
    "    .groupby('region', group_keys=False)\n",
    "    .apply(lambda x: x.nlargest(5, 'quality_score'))\n",
    "    .assign(\n",
    "        # 1. FASTA name per genome\n",
    "        sample_fasta=lambda d: (\n",
    "            d['region'].astype(str) + \"_\" +\n",
    "            d['name'].astype(str) + \"_\" +\n",
    "            d['accession_no'].astype(str)\n",
    "        ),\n",
    "\n",
    "        # 2. Sequence type\n",
    "        sequence_type=lambda d: np.where(\n",
    "            d['serovar'].str.lower().str.contains(\"enteritidis\"),\n",
    "            \"ST11\",\n",
    "            np.where(\n",
    "                d['serovar'].str.lower().str.contains(\"heidelberg\"),\n",
    "                \"ST15\",\n",
    "                np.nan\n",
    "            )\n",
    "        ),\n",
    "\n",
    "        # 3. FASTA assembly\n",
    "        fasta_source=lambda d: np.where(\n",
    "            d['accession_no'].str.startswith(\"GCF\"),\n",
    "            \"NCBI_RefSeq\",\n",
    "            np.where(\n",
    "                d['accession_no'].str.startswith(\"SRR\"),\n",
    "                \"EnteroBase_assembly\",\n",
    "                np.nan\n",
    "            )\n",
    "        )\n",
    "    )\n",
    "    [[\n",
    "        \"region\", \"name\", \"accession_no\", \"assembly_barcode\",\n",
    "        \"serovar\", \"sequence_type\",\n",
    "        \"source_niche\", \"source_details\",\n",
    "        \"collection_year\", \"country\", \"lab_contact\",\n",
    "        \"n50\", \"contig_number\", \"coverage\", \"quality_score\",\n",
    "        \"sample_fasta\", \"fasta_source\"\n",
    "    ]]\n",
    ")\n",
    "\n",
    "df_final['sample_fasta'] = df_final['sample_fasta'].str.replace(\" \", \"_\", regex=False)\n",
    "\n",
    "df_final[df_final['sample_fasta'].str.contains(\" \")]\n",
    "\n",
    "print(df_final)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8289bd19-547e-42fb-8a45-5fcc60247915",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_path = \"/salmonella-pangenome-SP/data/Metadata/metadata_samples.tsv\"\n",
    "\n",
    "df_final.to_csv(\n",
    "    out_path,\n",
    "    sep=\"\\t\",\n",
    "    index=False\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "82817060-2376-4e32-87a6-03c101186a82",
   "metadata": {},
   "outputs": [],
   "source": [
    "genomes_df = df_final[[\"sample_fasta\", \"accession_no\", \"fasta_source\"]]\n",
    "\n",
    "out_path = \"/salmonella-pangenome-SP/data/Metadata/genomes_manifest.tsv\"\n",
    "genomes_df.to_csv(\n",
    "    out_path,\n",
    "    sep=\"\\t\",\n",
    "    index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6d6a240-fbcf-4979-8766-ca0067faf4af",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
